### PR TITLE
cleanup tf unittests: part 2

### PR DESCRIPTION
**Pull request by stas00**
This is part 2 of https://github.com/huggingface/transformers/pull/6196 and the original issue https://github.com/huggingface/transformers/issues/5973. This PR brings the tf and pt tests and templates to an almost identical template.

**edit: as other parts of code have evolved, the part of this comment below this line is no longer relevant - see the development of this PR in the subsequent comments**

---

As discussed at https://github.com/huggingface/transformers/issues/5973#issuecomment-668113798, a wrapper was needed to turn plain `results` dicts into object-like dicts, so that the `results.key` works the same whether it's returned by a class or it was just a local dict. I'm not sure `DictAttr` is the best name, but I had to start somewhere - it'll be an easy thing to rename it once you help me to find a better name. I did look for a pre-existing library to facilitate this, but didn't find anything that doesn't require installing another package.

To remind - the key change in part1 and in this clean up is to replace:

```
-            result = {
-                "sequence_output": sequence_output.numpy(),
-                "pooled_output": pooled_output.numpy(),
-            }
-            self.parent.assertListEqual(
-                list(result["sequence_output"].shape), [self.batch_size, self.seq_length, self.hidden_size]
-            )
-            self.parent.assertListEqual(list(result["pooled_output"].shape), [self.batch_size, self.hidden_size])
```
with:
```
+            result = DictAttr({"sequence_output": sequence_output.numpy(), "pooled_output": pooled_output.numpy()})
+            self.parent.assertEqual(result.sequence_output.shape, (self.batch_size, self.seq_length, self.hidden_size))
+            self.parent.assertEqual(result.pooled_output.shape, (self.batch_size, self.hidden_size))
```
like it was done in part 1, except part 1 didn't need the wrapper - `results` were bona fide objects there.

there is also one torch test in this batch that required the wrapper too: `test_modeling_transfo_xl.py`

@sshleifer

### CODE REVIEW

**Comment by stas00** [comment]
Hmm, looks like `isort` and `flake` are in conflict with each other. `make style` reformatted the code to:
```
        result = DictAttr({"sequence_output": sequence_output.numpy(), "pooled_output": pooled_output.numpy(),})
```
and flake complains:
```
tests/test_modeling_tf_albert.py:139:110: E231 missing whitespace after ','
```
if I add space, isort removes it.

How do I resolve such conflict?

**Comment by stas00** [comment]
99% of this PR has been facilitated by running:

```
perl -0777 -pi -e 's#self.parent.assertListEqual\(
[\s\n]*
list\((result\w*)\[" ([^"]+) "\].(?:shape|size\(\))\),[\s\n]+\[ ( [^\]]* ) \],?
[\s\n]*
\)
#self.parent.assertEqual($1.$2.shape, ($3))#xmsg' tests/test*py templates/adding_a_new_model/tests/test_modeling_tf_xxx.py

perl -0777 -pi -e 's#
result\s=\s\{
([^}]+)
\}
#result = DictAttr({$1}) #xmsg' tests/test*py templates/adding_a_new_model/tests/test_modeling_tf_xxx.py

perl -pi -e 's|^(from transformers.testing_utils import .*?)$|$1, DictAttr|' tests/test_modeling_tf_albert.py tests/test_modeling_tf_bert.py tests/test_modeling_tf_ctrl.py tests/test_modeling_tf_distilbert.py tests/test_modeling_tf_electra.py tests/test_modeling_tf_flaubert.py tests/test_modeling_tf_gpt2.py tests/test_modeling_tf_mobilebert.py tests/test_modeling_tf_openai_gpt.py tests/test_modeling_tf_roberta.py tests/test_modeling_tf_t5.py tests/test_modeling_tf_transfo_xl.py tests/test_modeling_tf_xlm.py tests/test_modeling_tf_xlnet.py test_modeling_transfo_xl.py templates/adding_a_new_model/tests/test_modeling_tf_xxx.py

make style
# remove trailing comma that breaks flake
perl -pi -e 's|\),\}\)|)})|' tests/test_modeling_tf* templates/adding_a_new_model/tests/test_modeling_tf_xxx.py
```

**Comment by sgugger** [comment]
Hi @stas00. I'm in the process of changing all model outputs of TF models to the same as PyTorch and updated all the tests. This is in #6247 that I expect to merge today (once merge conflicts are handled and I've added the templates).

Could you wait a tiny bit and then apply the same PERL commands you did in #6196 ?

**Comment by stas00** [comment]
Oh, that's even better - yes, of course, I will wait, @sgugger! Thank you!

Could you also check on this one?
https://github.com/huggingface/transformers/blob/master/tests/test_modeling_transfo_xl.py#L94
It's the only remaining torch test with `results` dict.

**Comment by sgugger** [comment]
For the slow test, we can use `outputs1["last_hidden_state"]` and so forth in the asserts after. I went for the quickest fixes in the tests.

**Comment by stas00** [comment]
> For the slow test, we can use `outputs1["last_hidden_state"]` and so forth in the asserts after. I went for the quickest fixes in the tests.

Part of this rewrite was replacing `results["key"]` with `results.key` as specified. So this one would be different. All other tests have or will follow the latter style. There are two constructors in `test_modeling_transfo_xl.py` that do that.

**Comment by sgugger** [comment]
The outputs PR has been merged, so pinging you @stas00

**Comment by stas00** [comment]
I will rework this PR to match @sgugger changes.

**Comment by stas00** [comment]
@sgugger, after your merge, these 3 tf tests are still using the old style it seems:
```
FAILED tests/test_modeling_tf_transfo_xl.py::TFTransfoXLModelTest::test_transfo_xl_lm_head - AttributeError: 'dict' object has no attribute 'lm_logits_1'
FAILED tests/test_modeling_tf_transfo_xl.py::TFTransfoXLModelTest::test_transfo_xl_model - AttributeError: 'dict' object has no attribute 'hidden_states_1'
FAILED tests/test_modeling_tf_xlnet.py::TFXLNetModelTest::test_xlnet_lm_head - AttributeError: 'dict' object has no attribute 'all_logits_1'
```
I reverted to the `["key"]` style for these for now.

**Review by sgugger — APPROVED** [review]
Thanks for taking care of the merge conflicts. I guess we can deal with the 4 remaining tests (one PT, 3 TF) in a separate PR.

**Comment by stas00** [comment]
> I guess we can deal with the 4 remaining tests (one PT, 3 TF) in a separate PR.

Yes. Could you please instruct on how to proceed with those or will you take care of them?

**Comment by sgugger** [comment]
I think just using the result1/result2 in the assers instead of grouping everything in a dict should be fine.

File: tests/test_modeling_tf_transfo_xl.py
Comment (sgugger) [inline tests/test_modeling_tf_transfo_xl.py:102]: I don't think all the calls to numpy are necessary. Do the tests pass without them?

Comment (stas00) [reply tests/test_modeling_tf_transfo_xl.py]: It's not needed, indeed. removed and pushed.

**Comment by stas00** [comment]
please re-run CI - unrelated failure - thank you.

File: tests/test_modeling_tf_transfo_xl.py
Comment (sgugger) [inline tests/test_modeling_tf_transfo_xl.py:95]: Same for those here, I don't think you need that line and the equivalent for mems2

File: tests/test_modeling_tf_xlnet.py
Comment (sgugger) [inline tests/test_modeling_tf_xlnet.py:192]: Same remark for the .numpy in this file.

**Review by sgugger — APPROVED** [review]
Thanks @stas00 , there are a few .numpy left that we can clean up and this will be good to merge. Pinging @sshleifer or @LysandreJik for a second review.

File: tests/test_modeling_xlnet.py
Comment (sshleifer) [inline tests/test_modeling_xlnet.py:322]: can this be 
`self.assertListEqual([mem.size() for mem in result2.mems], [(self.seq_length, self.batch_size, self.hidden_size)]*self.num_hidden_layers)`

somewhat cleaner IMO but no big deal.

**Review by sshleifer — APPROVED** [review]

File: tests/test_modeling_xlnet.py
Comment (stas00) [reply tests/test_modeling_xlnet.py:322]: plus `.size()` => `.shape` - will do. thank you.

Comment (stas00) [reply tests/test_modeling_xlnet.py:322]: So changing:
```
-            list(list(mem.shape) for mem in mems_1),
-            [[self.mem_len, self.batch_size, self.hidden_size]] * self.num_hidden_layers,
+            [mem.shape for mem in mems_1],
+            [(self.mem_len, self.batch_size, self.hidden_size)] * self.num_hidden_layers,
```

Perl:

```
perl -0777 -pi -e 's#self.parent.assertListEqual\(
[\s\n]*
list \( list \( mem.(?:shape|size\(\))\)\sfor\smem\s(.*?) \),
[\s\n]*
\[\[ (.*?) \]\]\s\*\sself.num_hidden_layers,
#self.parent.assertListEqual(
            [mem.shape for mem $1],
            [($2)] * self.num_hidden_layers,#xmsg' tests/test*py
```

**Comment by stas00** [comment]
If there is any more cleanup you'd like to see send me the before and after code snippets.

**Review by sshleifer — APPROVED** [review]

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @stas00!

**Pull request merged**
